File: dynamic_programming/rithmstar.cpp
#include <iostream>

using namespace std;

int map[10002];
int dp[10001][7][7]={};
int panmove[7][7]={
        {},
        {0, 0, 1, 2, 1, 2, 3},
        {0, 1, 0, 1, 2, 1, 2},
        {0, 2, 1, 0, 3, 2, 1},
        {0, 1, 2, 3, 0, 1, 2},
        {0, 2, 1, 2, 1, 0, 1},
        {0, 3, 2, 1, 2, 1, 0}
};
int ans = 30001;

void findpath(int x, int y, int locate, int moving);
int main(){
    ios_base :: sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);

    int locate;
    cin >> locate;
    for(int i=1;i<=locate;i++){
        cin >> map[i];
    }
    for(int k=0;k<=locate;k++) {
        for (int i = 1; i < 7; i++) {
            for (int j = 1; j < 7; j++) {
                dp[k][i][j]=-1;
            }
        }
    }
    dp[0][1][3] = 0;
    for(int now=1;now<=locate;now++){
        for(int i=1;i<7;i++){
            for(int j=1;j<7;j++){
                if(dp[now-1][i][j]!=-1){
                    if(map[now]!=j){
                        int i_dp = dp[now-1][i][j] + panmove[i][map[now]];
                        if(dp[now][map[now]][j]!=-1){
                            dp[now][map[now]][j] = (dp[now][map[now]][j]>i_dp)?i_dp:dp[now][map[now]][j];
                        } else dp[now][map[now]][j] = i_dp;
                    }

                    if(map[now]!=i){
                        int j_dp = dp[now-1][i][j] + panmove[j][map[now]];
                        if(dp[now][i][map[now]]!=-1){
                            dp[now][i][map[now]] = (dp[now][i][map[now]]>j_dp)?j_dp:dp[now][i][map[now]];
                        } else dp[now][i][map[now]] = j_dp;
                    }

                }
            }
        }
    }
    int finali=0, finalj=0;
    for(int i=1;i<7;i++){
        for(int j=1;j<7;j++){
            if(dp[locate][i][j]!=-1){
                if(ans>dp[locate][i][j]){
                    ans = dp[locate][i][j];
                    finali=i;
                    finalj=j;
                }
            }
        }
    }
    findpath(finali, finalj, locate, ans);
    cout << "<" << finali << ", " << finalj << ">" << endl;
    return 0;
}

void findpath(int x, int y, int locate, int moving){
    if(x==y && y==0){
        cout << "ERROR";
        return;
    }
    if(locate==1){
        cout << "<" << x << ", " << y << "> -> ";
        return;
    }
    for(int i=1;i<7;i++){
        for(int j=1;j<7;j++){
            if(dp[locate-1][i][j]==-1) continue;
            int dmove = moving-dp[locate-1][i][j];
            if(dmove==panmove[i][map[locate]] || dmove==panmove[j][map[locate]]){
                findpath(i, j, locate-1, moving-dmove);
                if(moving!=ans) cout << "<" << x << ", " << y << "> -> ";
                return;
            }
        }
    }
}

File: dynamic_programming/rithmstar.py
class Rithmstar:
    panmove = [
        [],
        [0, 0, 1, 2, 1, 2, 3],
        [0, 1, 0, 1, 2, 1, 2],
        [0, 2, 1, 0, 3, 2, 1],
        [0, 1, 2, 3, 0, 1, 2],
        [0, 2, 1, 2, 1, 0, 1],
        [0, 3, 2, 1, 2, 1, 0]
    ]
    dp = []

    def __init__(self):
        self.cnt = int(input("노트 수를 입력하세요 : "))
        self.note = list(map(int, input("=> ").split()))
        self.note = [0] + self.note

        for i in range(self.cnt+1):
            n = []
            for j in range(7):
                n.append([-1, -1, -1, -1, -1, -1, -1])
            self.dp.append(n)
        self.dp[0][1][3] = 0
    
    def dp_cal(self):
        for now in range(1, self.cnt+1):
            for i in range(1, 7):
                for j in range(1, 7):
                    if self.dp[now-1][i][j] != -1:
                        if self.note[now]!=j:
                            i_dp = self.dp[now-1][i][j] + self.panmove[i][self.note[now]]
                            if self.dp[now][self.note[now]][j] != -1:
                                self.dp[now][self.note[now]][j] = min(self.dp[now][self.note[now]][j], i_dp)
                            else:
                                self.dp[now][self.note[now]][j] = i_dp
                        if self.note[now]!=i:
                            j_dp = self.dp[now-1][i][j] + self.panmove[j][self.note[now]]
                            if self.dp[now][i][self.note[now]] != -1:
                                self.dp[now][i][self.note[now]] = min(self.dp[now][i][self.note[now]],j_dp)
                            else:
                                self.dp[now][i][self.note[now]] = j_dp

    def find_path(self, x, y, locate, moving):
        if x==y and y==0:
            print("ERROR")
            return
        
        if locate==1:
            print("<1, 3> -> <{}, {}>".format(x, y), end="")
            return

        for i in range(1, 7):
            for j in range(1, 7):
                if self.dp[locate-1][i][j] == -1:
                    continue
                if i==x or j==y:
                    dmove = moving - self.dp[locate-1][i][j]
                    if dmove == self.panmove[i][self.note[locate]] or dmove == self.panmove[self.note[locate]][j]:
                        self.find_path(i, j, locate-1, moving-dmove)
                        print(" -> <{}, {}>".format(x, y), end="")
                        return


    def backtrack(self):
        ans = 30001
        finali = 0
        finalj = 0
        for i in range(1, 7):
            for j in range(1, 7):
                if self.dp[self.cnt][i][j] != -1:
                    if ans > self.dp[self.cnt][i][j]:
                        ans = self.dp[self.cnt][i][j]
                        finali = i
                        finalj = j
        self.find_path(finali, finalj, self.cnt, ans)
                    
my_game = Rithmstar()
my_game.dp_cal()
my_game.backtrack()